### PR TITLE
Create inventory peewee models

### PR DESCRIPTION
**Pull request by mc18g13**

### CODE REVIEW

**Review by HaGuesto — CHANGES_REQUESTED** [review]
### Functional review
I used the playground to test the queries and mutations.
- The `query_def` for base by `base_id` was wrong. It wanted a String not an Int. Already pushed an update. Also not your fault.
- The GraphQL query `allUsers` did not work since the name field in the model is not defined the right way. Already pushed a solution. 

- The GraphQL query `user(email)` does not work. I can not quickly solve it. This is the line and error where my debugger catches an error. Please resolve this.
image

- `createBox` mutation works.

All tests passed. That's nice. The errors above are integration errors with the mysql db which we do not test with pytest atm. Do you have any idea @mc18g13 how we could test that as well?

I also do not see any documentation how to write tests and create mock data :) would be nice.

File: flask/boxwise_flask/models/product.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+from boxwise_flask.db import db
+from boxwise_flask.models.base import Base
+from boxwise_flask.models.product_category import ProductCategory
+from boxwise_flask.models.product_gender import ProductGender
+from boxwise_flask.models.size_range import SizeRange
+from boxwise_flask.models.user import User
+
+
+class Product(db.Model):
+    amountneeded = FloatField(constraints=[SQL("DEFAULT 1")])
```
Comment (HaGuesto) [inline flask/boxwise_flask/models/product.py:19]: not needed anymore

File: flask/boxwise_flask/models/location.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+class Location(db.Model):
+    box_state = ForeignKeyField(
+        column_name="box_state_id",
+        constraints=[SQL("DEFAULT 1")],
+        field="id",
+        model=BoxState,
+        null=True,
+    )
+    base = ForeignKeyField(column_name="base_id", field="id", model=Base)
+    container_stock = IntegerField(constraints=[SQL("DEFAULT 0")])
```
Comment (HaGuesto) [inline flask/boxwise_flask/models/location.py:18]: please rename to `is_stockroom`

File: flask/boxwise_flask/models/product.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    )
+    modified = DateTimeField(null=True)
+    modified_by = ForeignKeyField(
+        column_name="modified_by", field="id", model=User, null=True,
+    )
+    name = CharField()
+    size_range = ForeignKeyField(
+        column_name="sizegroup_id", field="id", model=SizeRange, null=True
+    )
+    stockincontainer = IntegerField(constraints=[SQL("DEFAULT 0")])
```
Comment (HaGuesto) [inline flask/boxwise_flask/models/product.py:41]: Please rename to `in_shop`

File: flask/boxwise_flask/models/product.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+from boxwise_flask.models.product_category import ProductCategory
+from boxwise_flask.models.product_gender import ProductGender
+from boxwise_flask.models.size_range import SizeRange
+from boxwise_flask.models.user import User
+
+
+class Product(db.Model):
+    amountneeded = FloatField(constraints=[SQL("DEFAULT 1")])
+    base = ForeignKeyField(column_name="camp_id", field="id", model=Base, null=True)
+    category = ForeignKeyField(
```
Comment (HaGuesto) [inline flask/boxwise_flask/models/product.py:21]: This should probably be productcategory.

Comment (HaGuesto) [reply flask/boxwise_flask/models/product.py:21]: in the right convention

File: flask/boxwise_flask/models/user.py [hunk 5]
```diff
@@ -1,33 +1,62 @@
...
+        column_name="language", field="id", model=Language, null=True
+    )
     lastaction = DateTimeField()
+    lastlogin = DateTimeField()
+    modified = DateTimeField(null=True)
+    modified_by = ForeignKeyField(
+        column_name="modified_by", field="id", model="self", null=True,
+    )
+    name = CharField(column_name="naam", constraints=[SQL("DEFAULT ''")])
+    pass_ = CharField(column_name="pass", constraints=[SQL("DEFAULT ''")])
```
Comment (HaGuesto) [inline flask/boxwise_flask/models/user.py:37]: can u drop this one

File: flask/boxwise_flask/models/user.py [hunk 6]
```diff
@@ -1,33 +1,62 @@
...
+    resetpassword = CharField(null=True)
```
Comment (HaGuesto) [inline flask/boxwise_flask/models/user.py:38]: and this one too

File: flask/test/data/box.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
+from datetime import datetime
+
+import pytest
+from boxwise_flask.models.box import Box
+
+
+@pytest.fixture()
+def default_box(default_box_state, default_location):
+    mock_box = {
```
Comment (HaGuesto) [inline flask/test/data/box.py:9]: box always needs a product in it

File: flask/test/endpoint_tests/conftest.py [hunk 8]
```diff
@@ -1,12 +1,14 @@
 """Functionality for common test setups.
-See https://docs.pytest.org/en/stable/fixture.html#conftest-py-sharing-fixture-functions
+See https://docs import
+test.org/en/stable/fixture.html#conftest-py-sharing-fixture-functions
 
 In general, pass fixtures as arguments to a pytest test function in order to base the
```
Comment (HaGuesto) [inline flask/test/endpoint_tests/conftest.py:5]: beautiful. :)

File: flask/test/endpoint_tests/test_create_box.py [hunk 9]
```diff
@@ -19,23 +19,19 @@ def test_create_box(client):
                     created_by: "1"
                 }}"""
 
+    # TODO: add location, product and qr to the responses for this
```
Comment (HaGuesto) [inline flask/test/endpoint_tests/test_create_box.py:22]: What happened here?

File: flask/test/endpoint_tests/test_graphql.py [hunk 10]
```diff
@@ -38,17 +54,34 @@ def test_all_bases(client):
     for expected_base in bases:
         created_base = get_base_from_graphql(expected_base["id"], all_bases)
         assert created_base["id"] == expected_base["id"]
-        assert created_base["organisationId"] == expected_base["organisation_id"]
```
Comment (HaGuesto) [inline flask/test/endpoint_tests/test_graphql.py:41]: Why did you remove this assertion?

File: flask/test/endpoint_tests/test_user.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+    print(response_data.json["data"])
+    print(response_data.json["data"])
+    print(response_data.json["data"])
+    print(response_data.json["data"])
+    print(response_data.json["data"])
+    assert response_data.json["data"]["user"]["id"] == test_id
+
+
+# def test_all_users(client):
```
Comment (HaGuesto) [inline flask/test/endpoint_tests/test_user.py:72]: What happened here?

File: flask/test/endpoint_tests/test_user.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+                    lastlogin
+                    lastaction
+                }}
+            }}"""
+
+    data = {"query": graph_ql_query_string}
+    response_data = client.post("/graphql", json=data)
+    assert response_data.status_code == 200
+
+    print(response_data.json["data"])
```
Comment (HaGuesto) [inline flask/test/endpoint_tests/test_user.py:64]: What happened here? any reason for the copy/paste attack?

File: flask/test/model_tests/test_models.py [hunk 13]
```diff
@@ -2,13 +2,22 @@
...
 
 def test_model_method():
-    """Verify inserting test rows."""
 
-    Base.create(id=1, organisation_id=1, name="some text1", currency_name="hello")
-    Base.create(id=2, organisation_id=1, name="some text1", currency_name="hello")
-    Base.create(id=3, organisation_id=1, name="some text1", currency_name="hello")
-    Base.create(id=4, organisation_id=1, name="some text1", currency_name="hello")
-    Base.create(id=5, organisation_id=1, name="some text1", currency_name="hello")
+    Base.create(
```
Comment (HaGuesto) [inline flask/test/model_tests/test_models.py:6]: any reason why you do not use a fixture here?

File: flask/test/endpoint_tests/test_user.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    """Verify users GraphQL query endpoint"""
+
+    db.connect_db()
+    emails = [
+        "mr-anderson@matrix.co.uk",
+        "hamburgerman@beef.co.uk",
+        "marmalade@jam.co.uk",
+    ]
+
+    user = {
```
Comment (HaGuesto) [inline flask/test/endpoint_tests/test_user.py:17]: any reason why you do not use a fixture here?

File: flask/test/endpoint_tests/test_graphql.py [hunk 15]
```diff
@@ -38,17 +54,34 @@ def test_all_bases(client):
...
         assert created_base["currencyName"] == expected_base["currency_name"]
 
 
 def test_base(client):
     """Verify base GraphQL query endpoint"""
     bases = [
-        {"id": 1, "name": "oak-tree", "organisation_id": 1, "currency_name": "pound"},
-        {"id": 2, "name": "chicken", "organisation_id": 1, "currency_name": "peanuts"},
-        {"id": 3, "name": "sofa", "organisation_id": 1, "currency_name": "candles"},
+        {
```
Comment (HaGuesto) [inline flask/test/endpoint_tests/test_graphql.py:64]: any reason why you do not use a fixture here?

File: flask/test/endpoint_tests/test_graphql.py [hunk 16]
```diff
@@ -10,9 +9,27 @@ def get_base_from_graphql(id, base_query):
 def test_all_bases(client):
     """Verify allBases GraphQL query endpoint"""
     bases = [
-        {"id": 1, "name": "oak-tree", "organisation_id": 1, "currency_name": "pound"},
-        {"id": 2, "name": "chicken", "organisation_id": 1, "currency_name": "peanuts"},
-        {"id": 3, "name": "sofa", "organisation_id": 1, "currency_name": "candles"},
+        {
```
Comment (HaGuesto) [inline flask/test/endpoint_tests/test_graphql.py:12]: any reason why you do not use a fixture here?

**Review by HaGuesto — CHANGES_REQUESTED** [review]
###Code Review

Looks good. Thank you @mc18g13! The tests maybe need some clean-up or a reason why not :)

Most important thing for me would be to have a convention for the peewee class and field naming.

File: flask/boxwise_flask/models/base.py [hunk 17]
```diff
@@ -1,12 +1,48 @@
...
+from boxwise_flask.models.user import User
+from peewee import SQL, CharField, DateTimeField, ForeignKeyField, IntegerField
 
 
 class Base(db.Model):
-    organisation_id = IntegerField()
-    name = CharField()
-    currency_name = CharField(column_name="currencyname")
+    name = CharField(null=True)
+    currency_name = CharField(
```
[the remark this answers is not in the document]
Comment (mc18g13) [reply flask/boxwise_flask/models/base.py:11]: The convention used is that of python and is already defined by the rest of the code base and pre commit config right...?

[the remark this answers is not in the document]
Comment (HaGuesto) [reply flask/boxwise_flask/models/base.py:11]: I don't thinks so. I do not see a convention in these fields.

File: flask/boxwise_flask/models/user.py [hunk 18]
```diff
@@ -1,33 +1,62 @@
...
         user = User.select().where(User.email == email).get()
-        base_ids = UsergroupBaseAccess.get_all_base_id_for(user.usergroup_id)
+        base_ids = []
+        if user.usergroup:
+            base_ids = UsergroupBaseAccess.get_all_base_id_for_usergroup_id(
+                user.usergroup.id
+            )
         # base_ids is a peewee ModelSelect (so, many objects).
         # convert to dict 1 at a time,
         # and pull the base_id from that dict, and put in a list
```
Comment (pylipp) [inline flask/boxwise_flask/models/user.py:62]: I don't exactly understand how this works... the `User` model does not even have a `base_id`?

File: flask/test/data/base.py [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    bases_dict[mock_base["id"]] = mock_base
+    Base.create(**mock_base)
+
+    mock_base = {"id": 2, "seq": 1, "organisation_id": default_organisation["id"]}
+    bases_dict[mock_base["id"]] = mock_base
+    Base.create(**mock_base)
+
+    mock_base = {"id": 3, "seq": 1, "organisation_id": default_organisation["id"]}
+    bases_dict[mock_base["id"]] = mock_base
+    Base.create(**mock_base)
```
Comment (pylipp) [inline flask/test/data/base.py:25]: Can we use a loop instead? And write `bases` instead of `bases_dict` (I don't find it good style to include types in variable names)?
```
for i in range(1,4):
    mock_base = {"id": i, ...}
    bases[i] = mock_base
    Base.create(**mock_base)
```

File: flask/test/endpoint_tests/conftest.py [hunk 20]
```diff
@@ -1,12 +1,14 @@
 """Functionality for common test setups.
-See https://docs.pytest.org/en/stable/fixture.html#conftest-py-sharing-fixture-functions
+See https://docs import
+test.org/en/stable/fixture.html#conftest-py-sharing-fixture-functions
```
Comment (pylipp) [inline flask/test/endpoint_tests/conftest.py:3]: Is this on purpose?

File: flask/test/model_tests/test_base.py [hunk 21]
```diff
@@ -0,0 +1,38 @@
+import pytest
+from boxwise_flask.models.base import Base
+from playhouse.shortcuts import model_to_dict
+
+
+@pytest.mark.usefixtures("default_base")
+def test_get_base_from_id(default_base):
+    base = Base.get_from_id(default_base["id"])
+    base_dict = model_to_dict(base)
```
Comment (pylipp) [inline flask/test/model_tests/test_base.py:9]: You could use `base.id`, `base.seq` etc instead of converting to a dict. I find it'd highlight the difference between expected (dict) and actual (model instance) content

File: flask/test/model_tests/test_box.py [hunk 22]
```diff
@@ -1,31 +1,25 @@
...
-        print("input", new_box)
+    queried_box_dict = model_to_dict(queried_box)
+    if queried_box_dict != default_box:
+        print("queried_box ", queried_box_dict)
+        print("created_box ", default_box)
 
-    assert as_dict == new_box
+    assert queried_box_dict["id"] == default_box["id"]
+    assert queried_box_dict["box_id"] == default_box["box_id"]
+    assert queried_box_dict["box_state"]["id"] == default_box["box_state"]
```
Comment (pylipp) [inline flask/test/model_tests/test_box.py:18]: Is this the reason why `assert queried_box_dict == default_box` does not work?

**Review by pylipp — CHANGES_REQUESTED** [review]

**Comment by pylipp** [comment]
Congratz on the first PR for the new repo :tada: did you create the model definitions from pwiz?
Could not do the functional review yet.

**Comment by aerinsol** [comment]
Thanks for the excellent review Philipp! Yes, Matt created them from Pwiz.

File: flask/boxwise_flask/models/base.py [hunk 23]
```diff
@@ -1,12 +1,48 @@
...
+        column_name="currencyname", constraints=[SQL("DEFAULT 'Tokens'")]
```
[the remark this answers is not in the document]
Comment (mc18g13) [reply flask/boxwise_flask/models/base.py:12]: a lot of this is autogen from peewiz so its just straight from the db. I figured I'd try not to update the auto generated code too much in case we update the tables and have to gen things again.

File: flask/boxwise_flask/models/base.py [hunk 24]
```diff
@@ -24,13 +60,12 @@ def __str__(self):
 
     @staticmethod
     def get_all_bases():
-        return Base.select().order_by(Base.name)
+        return list(Base.select().order_by(Base.name))
```
[the remark this answers is not in the document]
Comment (mc18g13) [reply flask/boxwise_flask/models/base.py:63]: https://stackoverflow.com/questions/45501197/get-list-of-query-results-in-peewee

This sort of explains why you have list the first place, But I think the model classes should return the required data to the functions that uses it. rather than the functions having to understand how the model classes work.

File: flask/boxwise_flask/models/user.py (hunk 18 above)
Comment (mc18g13) [reply flask/boxwise_flask/models/user.py:62]: So this is a little confusing and I didn't write this originally :confused: . But because the user graphql query has a base id as part of the type def its appended to the object that returns in the case it exists. I shall move to the parent function call to make it more obvious how this works. it should also probably be camp_id. which will involved a frontend change as well :+1:

File: flask/test/data/box.py (hunk 7 above)
Comment (mc18g13) [reply flask/test/data/box.py:13]: why is product nullable in the database?

File: flask/test/endpoint_tests/conftest.py (hunk 20 above)
Comment (mc18g13) [reply flask/test/endpoint_tests/conftest.py:3]: the precommit config isnt working and complains about line width at the moment I was trying to get a commit in and shortening some of the lines. Not needed I'll revert

File: flask/test/model_tests/test_box.py (hunk 22 above)
Comment (mc18g13) [reply flask/test/model_tests/test_box.py:18]: assert queried_box_dict["box_state"]["id"] == default_box["box_state"]

File: flask/boxwise_flask/models/base.py [hunk 25]
```diff
@@ -35,13 +51,27 @@ class Base(db.Model):
...
         column_name="organisation_id", field="id", model=Organisation
     )
-    resettokens = IntegerField(constraints=[SQL("DEFAULT 0")], null=True)
-    schedulebreak = CharField(constraints=[SQL("DEFAULT '1'")])
-    schedulebreakduration = CharField(constraints=[SQL("DEFAULT '1'")])
-    schedulebreakstart = CharField(constraints=[SQL("DEFAULT '13:00'")])
-    schedulestart = CharField(constraints=[SQL("DEFAULT '11:00'")])
-    schedulestop = CharField(constraints=[SQL("DEFAULT '17:00'")])
-    scheduletimeslot = CharField(constraints=[SQL("DEFAULT '0.5'")])
+    resettokens = IntegerField(
```
Comment (pylipp) [inline flask/boxwise_flask/models/base.py:54]: reset_tokens?

**Review by pylipp — APPROVED** [review]

File: flask/boxwise_flask/models/user.py (hunk 18 above)
Comment (pylipp) [reply flask/boxwise_flask/models/user.py]: Ok nevermind :) maybe something for a tiny refactoring ticket rather @aerinsol

File: flask/boxwise_flask/models/qr_code.py [hunk 26]
```diff
@@ -0,0 +1,21 @@
+from peewee import SQL, CharField, DateTimeField, IntegerField
+
+from boxwise_flask.db import db
+
+
+class QRCode(db.Model):
```
[the remark this answers is not in the document]
Comment (pylipp) [reply flask/boxwise_flask/models/qr_code.py:5]: ping @HaGuesto

**Review by pylipp — APPROVED** [review]

File: flask/test/data/box.py (hunk 7 above)
Comment (aerinsol) [reply flask/test/data/box.py:13]: Yeah we should fix that in the database.

File: flask/boxwise_flask/models/user.py (hunk 18 above)
Comment (aerinsol) [reply flask/boxwise_flask/models/user.py]: That's weird. The user query has a base option? Not sure I understand how that happened, let's discuss on Slack

File: flask/boxwise_flask/models/qr_code.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (mc18g13) [reply flask/boxwise_flask/models/qr_code.py:5]: qrcode stands for Quick Response Code. I thought this fits in with the acronym

File: flask/test/data/base.py (hunk 19 above)
Comment (mc18g13) [reply flask/test/data/base.py]: I'd agree except when converting something of the same structure to a different type. makes it easier to understand the difference in two things that represent the same thing as seperate types.

File: flask/boxwise_flask/models/user.py [hunk 27]
```diff
@@ -1,36 +1,74 @@
...
-        return user
+    def get_from_email(email):
+        return (
+            User.select(User.id, User.name, User.email, User.usergroup)
+            .where(User.email == email)
+            .get()
+        )
+
+
+def get_user_from_email_with_base_ids(email):
```
Comment (HaGuesto) [inline flask/boxwise_flask/models/user.py:59]: Did you add that to resolve the error we had when using the graphql playground?

File: flask/boxwise_flask/models/usergroup_base_access.py [hunk 28]
```diff
@@ -17,7 +16,9 @@ def __str__(self):
         return self.name
 
     @staticmethod
-    def get_all_base_id_for(usergroup_id):
-        return UsergroupBaseAccess.select(UsergroupBaseAccess.base_id).where(
+    def get_all_base_id_for_usergroup_id(usergroup_id):
```
[the remark this answers is not in the document]
Comment (HaGuesto) [reply flask/boxwise_flask/models/usergroup_base_access.py:19]: no. access is given by base and module. This naming is fine.

**Review by HaGuesto — APPROVED** [review]
Looks good! 
GraphQL queries work, tests run through. 
Quite an effort, but major infra-structure is finally set up. Thanks @mc18g13

**Comment by HaGuesto** [comment]
nice i got the tests to fail. Stupid me

File: flask/boxwise_flask/models/user.py (hunk 27 above)
Comment (mc18g13) [reply flask/boxwise_flask/models/user.py:59]: I seperated it from the model to avoid confusion about the link between the base ids
